### PR TITLE
Improve CompilationContext.ContractName

### LINKED ISSUE #720: --contract-name option no longer working
PR #704 broke the `--contract-name` option. The point of the `--contract-name ` is to allow the build system to control the name of the output file.

### PR DESCRIPTION
**Pull request by erikzhang**
1. Revert #722
2. Fix #720

@devhawk @shargon

### CODE REVIEW

**Review by shargon — APPROVED** [review]

**Comment by devhawk** [comment]
As I understand it, we can't even run the tool until #721 is fixed, right? IMO, this PR (and all PRs frankly) for devpack should wait until #721 is fixed.

**Review by devhawk — CHANGES_REQUESTED** [review]
The more I think about this, the more I think that it's important NOT to tie the contract name and the base name for contract/manifest/debug info files

**Comment by erikzhang** [comment]
> The more I think about this, the more I think that it's important NOT to tie the contract name and the base name for contract/manifest/debug info files

The logic and behavior of this PR is the same as the current code, just easier to read.

**Comment by devhawk** [comment]
> > The more I think about this, the more I think that it's important NOT to tie the contract name and the base name for contract/manifest/debug info files
> 
> The logic and behavior of this PR is the same as the current code, just easier to read.

This is not accurate. `ProcessOutputs` currently uses `options.ContractName` as the base file name if it's not null and then falls back to using `context.ContractName`. That code was replaced with code that uses `context.ContractName` as the base file name.

Build tools need a mechanism to control the file name of the generated files, without depending on the contents of any source file. At the same time, I think it's important for developers to be able to manage the names of their contracts without constraints imposed by the build tools 

Maybe we should add a separate option to specify the base name of the generated files? The .NET C# compiler uses [`-out:`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/output#outputassembly) argument, but that would collide with the existing `--output` NCCS compiler argument. How about `--base-name`?

File: src/Neo.Compiler.CSharp/Program.cs [hunk 1]
```diff
@@ -126,27 +126,25 @@ private static int ProcessOutputs(Options options, string folder, CompilationCon
             if (context.Success)
             {
                 folder = options.Output ?? Path.Combine(folder, "bin", "sc");
-                var contractName = string.IsNullOrEmpty(options.ContractName)
-                    ? context.ContractName : options.ContractName;
                 Directory.CreateDirectory(folder);
-                string path = Path.Combine(folder, $"{contractName}.nef");
+                string path = Path.Combine(folder, $"{context.ContractName}.nef");
```
Comment (devhawk) [inline src/Neo.Compiler.CSharp/Program.cs:130]: Using context.ContractName here forces the contract name in the manifest to be the same as the base name of the generated files.

File: src/Neo.Compiler.CSharp/CompilationContext.cs [hunk 2]
```diff
@@ -52,7 +53,7 @@ public class CompilationContext
 
         public bool Success => diagnostics.All(p => p.Severity != DiagnosticSeverity.Error);
         public IReadOnlyList<Diagnostic> Diagnostics => diagnostics;
-        public string? ContractName { get; private set; }
+        public string? ContractName => Options.ContractName ?? displayName ?? assemblyName ?? className;
```
Comment (devhawk) [inline src/Neo.Compiler.CSharp/CompilationContext.cs:56]: This line overrides the contract name in source, re-breaking #674 and #696

**Comment by erikzhang** [comment]
> Maybe we should add a separate option to specify the base name of the generated files? The .NET C# compiler uses [-out:](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/output#outputassembly) argument, but that would collide with the existing --output NCCS compiler argument. How about --base-name?

I think it's okay. And `--contract-name` can be removed.

**Review by devhawk — APPROVED** [review]

**Pull request merged**